### PR TITLE
crimson/os/seastore: backref related clean ups

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 1]
```diff
@@ -685,23 +683,8 @@ SegmentCleaner::_retrieve_backref_extents(
     Cache::backref_extent_buf_entry_t::cmp_t> &&backref_extents,
   std::vector<CachedExtentRef> &extents)
 {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.cc:685]: _retrieve_backref_extents() can be removed

File: src/crimson/os/seastore/backref_manager.h [hunk 2]
```diff
@@ -97,6 +97,44 @@ class BackrefManager {
 					//   created by this insertion
   ) = 0;
 
+  virtual std::set<
+    backref_buf_entry_t,
+    backref_buf_entry_t::cmp_t> get_cached_backrefs_in_range(
```
Comment (cyx1231st) [inline src/crimson/os/seastore/backref_manager.h:88]: nit, add name alias to both `std::set<backref_buf_entry_t, backref_buf_entry_t::cmp_t>` and `std::set<Cache::backref_extent_buf_entry_t, Cache::backref_extent_buf_entry_t::cmp_t>`

File: src/crimson/os/seastore/segment_cleaner.cc [hunk 3]
```diff
@@ -761,90 +761,100 @@ SegmentCleaner::gc_reclaim_space_ret SegmentCleaner::gc_reclaim_space()
...
-      [this, &reclaimed, segment_id, &runs, end_paddr]() mutable {
-      reclaimed = 0;
-      runs++;
+    return ecb->with_transaction_intr(
+      Transaction::src_t::READ,
+      "get_backref_mappings",
+      [this, end_paddr](auto &t) {
+      return backref_manager.get_mappings(
+	t, *next_reclaim_pos, end_paddr
+      );
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_cleaner.cc:770]: IIUC this is under the assumption that all cleaner transactions -- trim_backref, clearner_trim, clean_reclaim -- will not run in parallel, so they won't conflict with each other in the current code.

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/os/seastore/segment_cleaner.cc (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.cc:770]: yes

File: src/crimson/os/seastore/segment_cleaner.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_cleaner.cc:685]: yeah

**Comment by xxhdx1985126** [comment]
Done

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
I think it should be good to merge after a rebase.

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by cyx1231st** [comment]
jenkins test api

**Pull request merged**